README.md
# [`gocd_agent`](https://github.com/JamesMGreene/chef-gocd-agent) [![GitHub Latest Release](https://badge.fury.io/gh/JamesMGreene%2Fchef-gocd-agent.png)](https://github.com/JamesMGreene/chef-gocd-agent) [![Build Status](https://secure.travis-ci.org/JamesMGreene/chef-gocd-agent.png?branch=master)](https://travis-ci.org/JamesMGreene/chef-gocd-agent) [![Chef Cookbook](http://img.shields.io/cookbook/v/gocd_agent.svg)](https://supermarket.chef.io/cookbooks/gocd_agent)

A [Chef](https://www.chef.io/chef/) Cookbook to install and configure a [ThoughtWorks Go CD (GoCD)](http://www.go.cd/) Agent/Client.


## Supported Platforms

### Verified

 - **RHEL:** Amazon Linux


### Unverified

 - **RHEL:** RedHat, CentOS, Scientific, Oracle, Fedora, etc.
 - **Debian:** Debian, Ubuntu, LinuxMint, etc.
 - **MacOS X:** MacOS X, MacOS X Server, etc.
 - **Windows:** Windows, MinGW32, MSWin, etc.
 - **Solaris:** Solaris, Solaris2, OpenSolaris, OmniOS, SmartOS, etc.



## Requirements

### Other Cookbooks
- [`java`](https://supermarket.chef.io/cookbooks/java) - GoCD Agents require that OpenJDK/Java 7 is installed.


### `package`s
- `unzip` - In some logic branches, the `unzip` utility method will be used to decompress an archive file.



## Attributes

### Primary

```chef
# The Basics
default[:gocd_agent][:name]    = 'go-agent'
default[:gocd_agent][:version] = '15.2.0'
default[:gocd_agent][:release] = '2248'

# Configure communication with the GoCD Server
default[:gocd_agent][:gocd_server][:host] = '127.0.0.1'
default[:gocd_agent][:gocd_server][:port] = 8153

# Auto-register new GoCD Agents with the GoCD Server?
default[:gocd_agent][:auto_register][:key]          = nil
default[:gocd_agent][:auto_register][:resources]    = []
default[:gocd_agent][:auto_register][:environments] = []
# Requires GoCD 15.2.0 or higher
default[:gocd_agent][:auto_register][:hostname]     = node[:hostname]
```


### Secondary

There are many other attributes specified for this cookbook.  To see them in full, review the code: [attributes/default.rb](https://github.com/JamesMGreene/chef-gocd-agent/blob/master/attributes/default.rb)



## Usage

Include the `gocd_agent` default recipe to install a GoCD Agent/Client on your system based on the default installation method:

```chef
include_recipe 'gocd_agent'
```

**Synonyms:** `gocd_agent::default`, `gocd_agent::install`


### Directed Install Methods

#### Package

Install the GoCD Agent from official packages [but _without_ using repos]:

```chef
# Not necessary to set because it's the default, when available on a platform
node[:gocd_agent][:install_method] = 'package'
include_recipe 'gocd_agent'

# OR:

include_recipe 'gocd_agent::install_from_package'
```


#### Binary

Install the GoCD Agent from official prebuilt binaries:

```chef
# Not necessary to set because it's the default, when available on a platform
node[:gocd_agent][:install_method] = 'binary'
include_recipe 'gocd_agent'

# OR:

include_recipe 'gocd_agent::install_from_binary'
```


#### Source

Install the GoCD Agent from the released source archive:

```chef
# Not necessary to set because it's the default, when nothing else is available on a platform
node[:gocd_agent][:install_method] = 'source'
include_recipe 'gocd_agent'

# OR:

include_recipe 'gocd_agent::install_from_source'
```



## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Added some feature'`)
4. Run the unit tests (`bundle exec rake spec`)
5. Run test kitchen (`bundle exec kitchen test`)
6. Push to the branch (`git push origin my-new-feature`)
7. Create new Pull Request



## License

Copyright (c) 2015, James M. Greene (MIT License)


CHANGELOG.md
# `gocd_agent` CHANGELOG

This file is used to list changes made in each version of the `gocd_agent` cookbook.


## `1.1.11`
- Dropped the `:user` attribute from the `Mixlib::ShellOut` command when running on Windows.

## `1.1.10`
- Relaxed version constraint for the `java` dependency cookbook.

## `1.1.9`
- Now setting the `USER` environment variable on Linux.

## `1.1.8`
- Adjusted the `service` resource usage to stop the service before installing so that it is effectively restarted if the service was indeed already installed.

## `1.1.7`
- Added support for setting the `HOME` environment variable on Linux via new attribute `node['gocd_agent']['home']`.

## `1.1.6`
- Added support for prepending to the `PATH` environment variable via new attribute `node['gocd_agent']['path']`.

## `1.1.5`
- Fixed some instances of using non-existent String method 'trim' instead of 'strip'.

## `1.1.4`
- Drop the newer `:login` attribute from use of `Mixlib::ShellOut`... too modern for OpsWorks's Chef 11.10's "mixlib-shellout-1.3"!

## `1.1.3`
- Added better internal testing for Travis CI with `knife cookbook test`

## `1.1.2`
- Switched from using backtick-escaped line scripts to `Mixlib::ShellOut`

## `1.1.1`
- Added default value searches for `node['gocd_agent']['java_home']`.

## `1.1.0`
- Added new attribute to allow prereqs to **NOT** be forcibly installed.

## `1.0.12`
- Modified template for 'autoregister.properties' file to remove excess whitespace.

## `1.0.11`
- Adding attribute nodes for setting `work_dir` and `java_home` on Linux.

## `1.0.10`
- Fixed another attribute reference issue during Chef Compile phase vs. Converge phase, this time in
  the `service` resource.

## `1.0.9`
- Removed unnecessary `version` attribute from `package` installer resource.

## `1.0.8`
- Fixed issue with `node['gocd_agent']['installation_source']` being empty for installers due to
  their attributes being evaluated during the Chef Compile phase instead of the Converge phase by
  using `lazy` attribute evaluation where needed.
- Removed debugging code.

## `1.0.7`
- Syntax error fix. Sorry!

## `1.0.6`
- MOAR DEBUGGING!

## `1.0.5`
- Added some debugging code into a private recipe for analysis.

## `1.0.4`
- Shuffled some attribute definitions around to get the correct values propagated.

## `1.0.3`
- Added a link to the `java` cookbook dependency

## `1.0.2`
- Initial release of the `gocd_agent` cookbook!

- - -
Check the [Markdown Syntax Guide](http://daringfireball.net/projects/markdown/syntax) for help with Markdown.

The [Github Flavored Markdown page](http://github.github.com/github-flavored-markdown/) describes the differences between markdown on github and standard markdown.


attributes/default.rb
###
# Do NOT use this file to override this cookbook's default attributes.
# Instead, please use the "customize.rb" attributes file, which will keep
# your adjustments separate from the codebase and make it easier to upgrade.
#
# However, you should not edit "customize.rb" directly. Instead, create a
# "{thisCookbookName}/attributes/customize.rb" file in your own cookbook
# repository and put the overrides in YOUR "customize.rb" file.
#
# Do NOT create an "{thisCookbookName}/attributes/default.rb" in your
# cookbooks. Doing so would completely override this file and might cause
# upgrade issues.
###

# The Basics
default['gocd_agent']['name']    = 'go-agent'
default['gocd_agent']['version'] = '15.2.0'
default['gocd_agent']['release'] = '2248'

# Configure communication with the GoCD Server
default['gocd_agent']['gocd_server']['host'] = '127.0.0.1'
default['gocd_agent']['gocd_server']['port'] = 8153

# Configure local agent settings; ideally, you should leave these unset/default
default['gocd_agent']['work_dir'] = nil
default['gocd_agent']['home'] = '/var/go'
default['gocd_agent']['path'] = ENV['PATH']
if !(ENV['JAVA_HOME'].nil? || ENV['JAVA_HOME'].empty?)
  default['gocd_agent']['java_home'] = ENV['JAVA_HOME']
else
  case node['platform_family']
  when 'windows'
    cmd = Mixlib::ShellOut.new('echo "${JAVA_HOME}"')
  else
    cmd = Mixlib::ShellOut.new('echo "${JAVA_HOME}"', :user => 'root')
  end
  cmd.run_command
  if !(cmd.error? || cmd.stdout.nil? || cmd.stdout.strip.empty?)
    default['gocd_agent']['java_home'] = cmd.stdout.strip
  end
end



# Auto-register new GoCD Agents with the GoCD Server?
default['gocd_agent']['auto_register']['key']          = nil
default['gocd_agent']['auto_register']['resources']    = []
default['gocd_agent']['auto_register']['environments'] = []
# Requires GoCD 15.2.0 or higher
default['gocd_agent']['auto_register']['hostname']     = node['hostname']

# User/Group
default['gocd_agent']['user']  = 'go'
default['gocd_agent']['group'] = 'go'

# Install prereqs?
default['gocd_agent']['install_prereqs'] = true


# Configure the service name
case node['platform_family']
when 'solaris', 'solaris2', 'opensolaris', 'omnios', 'smartos'
  default['gocd_agent']['service'] = 'go/agent'
when 'windows'
  default['gocd_agent']['service'] = 'Go Agent'
else
  default['gocd_agent']['service'] = 'go-agent'
end



case node['platform_family']
when 'rhel', 'fedora'
  # yum (rpm)
  default['gocd_agent']['install_method'] = 'package'

when 'debian'
  # apt/apt-get (deb)
  default['gocd_agent']['install_method'] = 'package'

when 'solaris', 'solaris2', 'opensolaris', 'omnios', 'smartos'
  # pkgadd/pkgin (pkgsrc) - gzipped
  default['gocd_agent']['install_method'] = 'package'

when 'mac_os_x'
  # app - zipped
  default['gocd_agent']['install_method'] = 'binary'

when 'windows'
  # exe - Setup Installer
  default['gocd_agent']['install_method'] = 'binary'

else
  # Other platforms: primarily other Linux/UNIX/BSD variants
  # source - zipped
  default['gocd_agent']['install_method'] = 'source'

end


case node['gocd_agent']['install_method']
when 'package'

  case node['platform_family']
  when 'rhel', 'fedora'
    # yum (rpm)
    default['gocd_agent']['download_dir'] = '/gocd/gocd-rpm'
    default['gocd_agent']['download_file_extension'] = '.noarch.rpm'
    default['gocd_agent']['decompress'] = nil

  when 'debian'
    # apt/apt-get (deb)
    default['gocd_agent']['download_dir'] = '/gocd/gocd-deb'
    default['gocd_agent']['download_file_extension'] = '.deb'
    default['gocd_agent']['decompress'] = nil

  when 'solaris', 'solaris2', 'opensolaris', 'omnios', 'smartos'
    # pkgadd/pkgin (pkgsrc) - gzipped
    default['gocd_agent']['download_dir'] = '/gocd/gocd'
    default['gocd_agent']['download_file_extension'] = '-solaris.gz'
    default['gocd_agent']['decompress'] = 'gzip'

  end

when 'binary'

  case node['platform_family']
  when 'mac_os_x'
    # app - zipped
    default['gocd_agent']['download_dir'] = '/gocd/gocd'
    default['gocd_agent']['download_file_extension'] = '-osx.zip'
    default['gocd_agent']['decompress'] = 'zip'

  when 'windows'
    # exe - Setup Installer
    default['gocd_agent']['download_dir'] = '/gocd/gocd'
    default['gocd_agent']['download_file_extension'] = '-setup.exe'
    default['gocd_agent']['decompress'] = nil

  end

when 'source'

  # source - zipped
  default['gocd_agent']['download_dir'] = '/gocd/gocd'
  default['gocd_agent']['download_file_extension'] = '.zip'
  default['gocd_agent']['decompress'] = 'zip'

else
  raise "Unexpected value for `node['gocd_agent']['install_method']`: #{node['gocd_agent']['install_method']}"
end


# Configure the download
default['gocd_agent']['download_host'] = 'http://dl.bintray.com'
default['gocd_agent']['download_file_basename'] = "#{node['gocd_agent']['name']}-#{node['gocd_agent']['version']}-#{node['gocd_agent']['release']}"
default['gocd_agent']['download_file'] = "#{node['gocd_agent']['download_file_basename']}#{node['gocd_agent']['download_file_extension']}"
default['gocd_agent']['download_url']  = "#{node['gocd_agent']['download_host']}#{node['gocd_agent']['download_dir']}/#{node['gocd_agent']['download_file']}"


# Import custom attribute overrides
include_attribute 'gocd_agent::customize'
